politeiawww/politeiawww.go

```diff
@@ -859,6 +867,100 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
 	}
 }
 
+// updateBestBlock updates the cached best block.
+func (p *politeiawww) updateBestBlock(bestBlock uint64) {
+	p.bbMtx.Lock()
+	defer p.bbMtx.Unlock()
+	p.bestBlock = bestBlock
+}
+
+// getBestBlock returns the cached best block if there is an active websocket
+// connection to dcrdata. Otherwise, it requests the best block from politeiad
+// using the the decred plugin best block command.
+func (p *politeiawww) getBestBlock() (uint64, error) {
+	p.bbMtx.RLock()
+	bb := p.bestBlock
+	p.bbMtx.RUnlock()
+
+	// the cached best block will equal 0 if there is no active websocket
+	// connection to dcrdata, or if no new block messages have been received
+	// since a connection was established.
+	if bb == 0 {
+		return p.getBestBlockDecredPlugin()
+	}
+
+	return bb, nil
+}
+
+// resetPiDcrdataWSSubs is responsible for resetting the wsDcrdata connection
+// and making necessary changes to the required changes to the politeiawww
+// state so that the service continues to function properly during and after
+// the reconnection.
+func (p *politeiawww) resetPiDcrdataWSSubs() error {
+	// The cached best block is set to zero so that in the time between
+	// reconnection and receiving the first new block message, instead of
+	// using the old cached value, politeiad is queried for the best block.
+	p.updateBestBlock(0)
+
+	return p.wsDcrdata.reconnect()
+}
+
+// setupPiDcrdataWSSubs subscribes and listens to websocket messages from
+// dcrdata that are needed for pi.
+func (p *politeiawww) setupPiDcrdataWSSubs() error {
+	err := p.wsDcrdata.subscribe(newBlockSub)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+	listenerLoop:
+		for {
+			msg, ok := <-p.wsDcrdata.client.Receive()
+			if !ok || msg == nil {
+				// After closing wsDcrdata, the channel channel will send a nil
+				// message. This check is just here to avoid a spew of
+				// unnecessary error messages.
+				if p.wsDcrdata.isShutdown() {
+					break listenerLoop
+				}
+
+				log.Errorf("Error receiving msg from dcrdata, msg: %v", msg)
+				err = p.resetPiDcrdataWSSubs()
+				if err != nil {
+					log.Errorf("Unable to reconnect to dcrdata, no longer " +
+						"listening.")
+					break listenerLoop
+				}
+				continue
+			}
+
+			switch m := msg.Message.(type) {
+			case *exptypes.WebsocketBlock:
+				log.Debugf("wsDcrdata message WebsocketBlock(height=%s)",
+					m.Block.Height)
+				p.updateBestBlock(uint64(m.Block.Height))
+			case *pstypes.HangUp:
+				log.Infof("Dcrdata has hung up. Will reconnect...")
+				err = p.resetPiDcrdataWSSubs()
+				if err != nil {
+					log.Errorf("Unable to reconnect to dcrdata, no longer " +
+						"listening.")
+					break listenerLoop
+				}
+				log.Infof("Successfully reconnected to dcrdata")
+			case int:
+				// Ping messages are of type int
```

Comment:
Glad you found the right type for pings.  Sorry I said it was pstypes... type.  The type really should be made specific in a future psclient and pubsub/types.

Reply:
It's all good, but yes that would definitely be nicer.